### PR TITLE
feat(i18n/l10n): add crowdin script + integration into CI

### PR DESCRIPTION
**Pull request by jmfrancois**
**What is the problem this PR is trying to solve?**

We don't want to store translations into UI repository
We don't have an API to store them internally
We have some hard coded translation into components/stories

**What is the chosen solution to this problem?**

Let's try crowdin service.
Create a script with the following features:

* status: give a status of translations
* download: download all existing translations
* upload: upload the result of i18next extraction

Plug the upload to the CI when running on master (after each PR merged)

The main idea is to trigger the download with that script during the release process so the locales are available under `@talend/react-components` **/locales** and `@talend/react-forms`**/locales**

You can try with @talend/react-components and @talend/react-forms using the following version: **0.83.0-beta-2**

We have moved the parse-i18n into *prepublish* of each packages.

**Please check if the PR fulfills these requirements**

* [x] The PR commit message follows our [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md)
* [ ] Tests for the changes have been added (for bug fixes / features) And [non reg](./screenshots.md) done before need review
* [ ] Docs have been added / updated (for bug fixes / features)
* [ ] Related design / discussions / pages (not in jira), if any, are all linked or available in the PR

**[ ] This PR introduces a breaking change**

### CODE REVIEW

File: crowdin.js [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+function getDestination(fileEntry) {
+	debug(fileEntry.toString());
+	const path = fileEntry.entryName;
+	if (fileEntry.name === 'tui-components.json') {
+		return `packages/components/locales/${path}`; // "ja/tui-components.json"
+	} else if (fileEntry.name === 'tui-forms.json') {
+		return `packages/forms/locales/${path}`; // "ja/tui-forms.json"
+	} else if (fileEntry.isDirectory) {
+		debug(`directory found supposed to be a language folder: ${path}`);
+		if (!fs.existsSync(`packages/components/locales/${path}`)) {
```
Comment (jsomsanith) [inline crowdin.js]: `getDestination()` should not create anything, but just return a destination.
Perhaps you can move that into `extractAndSave()`. At the beginning, test if it's a folder, if so create the folders, else write.

File: crowdin.js [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+function debug(...args) {
+	if (program.verbose) {
+		console.log(...args);
+	}
+}
+
+// TODO: prompt for the PROJECT_KEY
+
+const PROJECT_KEY = program.key;
```
Comment (jsomsanith) [inline crowdin.js]: Can you rename it to `API_KEY` ?

File: crowdin.js [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+}
+
+function handleDownloadRequest(error, response, body) {
+	onError(error);
+	const zip = new AdmZip(body);
+	zip.getEntries().forEach(extractAndSave);
+}
+
+function cleanUpBeforeDownload() {
+	if (!fs.existsSync('packages/components/locales')) {
```
Comment (jsomsanith) [inline crowdin.js]: Shouldn't we remove all the folder content ?

File: crowdin.js [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+		},
+		encoding: null,
+	};
+	debug(`download url OPTIONS ${options}`);
+	request(options, handleDownloadRequest);
+}
+
+function upload() {
+	console.log('please use curl');
+	const files = FILES.map(file => `\n  -F "files[/${file.name}]=@${file.path}"`);
```
Comment (jsomsanith) [inline crowdin.js]: Could we run `yarn extract-i18n` so we don't have to do it before running the script ?

File: crowdin.js [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+}
+
+function getDestination(fileEntry) {
+	debug(fileEntry.toString());
+	const path = fileEntry.entryName;
+	if (fileEntry.name === 'tui-components.json') {
+		return `packages/components/locales/${path}`; // "ja/tui-components.json"
+	} else if (fileEntry.name === 'tui-forms.json') {
+		return `packages/forms/locales/${path}`; // "ja/tui-forms.json"
+	} else if (fileEntry.isDirectory) {
```
[the remark this answers is not in the document]
Comment (jmfrancois) [reply crowdin.js]: AFAIK containers has no translations. I see i18next call but no locales extracted

File: crowdin.js (hunk 3 above)
Comment (jmfrancois) [reply crowdin.js]: I don't think this is needed.
I have renamed the function ensureLocalesExists

File: crowdin.js (hunk 4 above)
Comment (jmfrancois) [reply crowdin.js]: Good idea, just switch it. I have added the call to crowdin download in the extract

File: crowdin.js [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+const URL = `https://api.crowdin.com/api/project/${PROJECT_ID}`;
+
+if (!API_KEY) {
+	throw new Error('you must set -k or put a key in .crowdin config file');
+}
+
+const FILES = [
+	{
+		name: 'tui-components.json',
+		path: `${__dirname}/i18n/components/en/tui-components.json`,
```
Comment (jare-talend) [inline crowdin.js]: we can store the template on each packages in packages/{{package}}/locales/template/{{namespace}}.

So, we can add some variants for a key which are not parsed by i18next-parser (plural, context, ...)

File: packages/components/stories/config/i18n.js [hunk 7]
```diff
@@ -1,67 +1,31 @@
...
-			},
-		},
-	},
-	debug: false,
-	wait: true, // globally set to wait for loaded translations in translate hoc
-});
+let i18nInitialized = false;
+const languages = ['en', 'fr', 'ja'];
+
+if (process.env.NODE_ENV === 'production') {
```
Comment (jare-talend) [inline packages/components/stories/config/i18n.js]: you can remove this. On storybook, you doesn't need to check the env.

File: packages/components/.gitignore [hunk 8]
```diff
@@ -41,5 +41,7 @@ hs_err_pid*
 # Talend web apps build directory
 build/
 lib/
-
+locales/fr
+locales/en
+locales/ja
```
[the remark this answers is not in the document]
Comment (jmfrancois) [reply packages/components/.gitignore]: Yes I'm not proud of this one but I want to save the template; but not the locales

**Comment by jmfrancois** [comment]
Crowdin doesn't cover the need

**Comment by jare-talend** [comment]
can you explain why ?

**Pull request closed without merging**